python/ray/data/_internal/batcher.py

```diff
@@ -78,3 +121,203 @@ def next_batch(self) -> Block:
+    #  1. Build added blocks into a concrete shuffle buffer.
+    #  2. Generate random shuffle indices.
+    # Note that (1) and (2) only happen when new blocks are added, upon the next
+    # retrieval. I.e., if no new blocks have been added since the last batch retrieval,
+    # and there are still batches in the existing concrete shuffle buffer to be yielded,
+    # then each batch retreival will only involve slicing the batch out of the concrete
+    # shuffle buffer.
+    #
+    # Similarly, adding blocks is very cheap. Each added block will be appended to a
+    # list, with concatenation of the underlying data delayed until the next batch
+    # retrieval.
+    #
+    # Since (1) runs of block additions are cheap, and (2) runs of batch retrievals are
+    # cheap, callers of ShufflingBatcher are encouraged to add as many blocks as
+    # possible (up to the shuffle buffer capacity), followed by retrieving as many
+    # batches as possible (down to the shuffle buffer minimum size), in such contiguous
+    # runs.
+
+    def __init__(
+        self,
+        batch_size: Optional[int],
+        shuffle_buffer_capacity: Optional[int] = None,
+        shuffle_buffer_min_size: Optional[int] = None,
+        shuffle_seed: Optional[int] = None,
+    ):
+        """Constructs a random-shuffling block batcher.
+
+        Args:
+            batch_size: Record batch size.
+            shuffle_buffer_capacity: Soft maximum number of rows allowed in the local
+                in-memory shuffle buffer. This must be greater than or equal to
+                ``batch_size``. Note that this is a soft max: if the buffer is currently
+                smaller than this max, we will add a new data block to the buffer, but
+                this new data block may push the buffer over this max; we don't take the
+                size of the new data block into account when doing this capacity check.
+                Default is ``max(2 * shuffle_buffer_min_size, shuffle_buffer_min_size +
+                batch_size)`` if ``shuffle_buffer_min_size`` is given, otherwise the
+                default is ``10 * batch_size``.
+            shuffle_buffer_min_size: Minimum number of rows that must be in the local
+                in-memory shuffle buffer in order to yield a batch. This must be greater
+                than or equal to ``batch_size`` and must be less than
+                ``shuffle_buffer_capacity``. Increasing this will improve the randomness
+                of the shuffle but may increase the latency to the first batch.
+                Default is
+                ``max(min(shuffle_buffer_capacity // 2, shuffle_buffer_capacity -
+                batch_size), batch_size)``.
+            shuffle_seed: The seed to use for the local random shuffle.
+        """
+        if batch_size is None:
+            raise ValueError("Must specify a batch_size if using a local shuffle.")
```

Comment:
We could look at a separate, simple local shuffling option for the `batch_size=None` case, e.g. an unbounded local shuffler that fetches a block, shuffles it, and yields the shuffled block. Supporting local shuffling for the `batch_size=None` case is not a P0, so we don't allow it for now.